types/wlr_surface.c

```diff
@@ -169,10 +169,16 @@ static void surface_state_viewport_src_size(struct wlr_surface_state *state,
 
 static void surface_state_finalize(struct wlr_surface *surface,
 		struct wlr_surface_state *state) {
+	if (state->buffer_resource) {
+		state->buffer = wlr_buffer_import(surface->renderer,
+				state->buffer_resource);
+		wlr_buffer_lock(state->buffer);
```

Comment:
`wlr_buffer_import` returns a buffer which is already locked. No need to lock it a second time.

Reply:
I'm hitting an assert without it, tracked it down to this one yesterday.

> Assertion failed: buffer->n_locks > 0 (types/wlr_buffer.c: wlr_buffer_unlock: 58)

Reply:
In `surface_state_copy` the previous version was copying the pointer to the `wlr_buffer` without locking (`next->buffer = state->buffer` should have been `next->buffer = wlr_buffer_lock(state->buffer)`). I didn't mention it because I suggested to remove that code.

tl;dr I think it should work better with the new version of the PR. Can you try again removing it?

Reply:
Yup, it works